Massive MC and subsystem rewrite

MrsTonedOne opened the pull request:
fixes #9072
mostly fixes #9031 (there is another issue report up for what this doesn't fix)
fixes #4830 (pipenet's init running in the background was causing a race condition causing some pipenets to fail to setup fully. this would not be noticed until something new attempted to add itself to the bad pipenet)
Changes:
## MC:

No longer tracks a subsystem's cpu usage. This was basically worthless and took up space on the stat panel

Can calculate wait down to a tenth of a decisecond to make it fps/world.ticklag agnostic

Now allows subsystems to have a dynamic wait, this is based on a ratio of how long that subsystem has been taking to process(cost). (This system allows for an upper and lower bounds, and a changeable cost delta for each subsystem) (The algorithm is just cost*cost_delta, defaults to a cost delta of 3)

MC can now be told to init a zlevel
## All Subsystems:

Stats panel now allows child subsystems to pass it a message to add to its stats entry. All subsystems have been moved over to this system - This should cut down on subsystems having to copy and paste the stats proc in order to add to it

All subsystems now properlly handle being given a zlevel in their init proc
## Subsystem changes:
### Air:

Added air to the dynamic wait subsystem. upper bound: 50, lower bound: 5, cost delta: 3 times process cost

Air now fires 4 times faster when it can do so without lagging things up

Pipenet subsystem has been merged into air

Atmos machinery now processes with process_atmos(), ticked by air, not machinery.

Hotspots (the fire object) are now object pooled
### Pipenet:

Deleted, added to air
### Machinery:

Moved all atmos calcualtions in all objects's process() to process_atmos().
### Lighting:

Added Lighting to the dynamic wait subsystem. upper bound: 20, lower bound: 5, cost delta: 3 times process cost
### Ticker:

~~Fixed ticker not updating the lobby panel when game start delayed~~ #9228 fixed this before me

~~Fixed the game start timer updating rapidly from queued fires when game start delay is removed~~ #9228 fixed this before me

Fixed start now not causing a tip of the round to play.
### Garbage/qdel:

qdel will now limit its process time to 2ds a fire.

qdel can now be given hints as a return to Destroy() as to what should be done with the object.
the options are:
- queue: (default) this is the normal behavior.
- letmelive: old default to non-null/zero. does nothing with the object
- iwillgc: functionally the same as above, mainly to let people working with objects know that the object will not be queued for GC checking
- harddel: this will queue the object to be deleted without storing a soft reference, mainly to save locate() processing time.
- harddel_now: this will del() the object. To allow for a clean removal of every del() in the code thats not in qdel

All objects have been updated to the new system, harddel and iwillgc was not added to any objects.

Fixed some objects not GCing because they didn't properlly clear references in Destory()

Fixed some objects getting qdel'ed preventing other objects from getting GCed because they did not null their reference to that object.

MrsTonedOne committed:
Massive MC and subsystem rewrite
MC:
	No longer tracks a subsystem's cpu usage. This was basically worthless and took up space on the stat panel
	Can calculate wait down to a tenth of a decisecond to make it fps/world.ticklag agnostic
	Now allows subsystems to have a dynamic wait, that is based on a ratio of how long that subsystem has been taking to process(cost). (This system allows for upper and lower bounds, and an changeable cost delta for each subsystem)
	MC can now be told to init a zlevel

All Subsystems:
	Stats panel now allows child subsystems to pass it a message to add to its stats entry. All subsystems have been moved over to this system - This should cut down on subsystems having to copy and paste the stats proc in order to add to it
	All subsystems now properlly handle being given a zlevel in their init proc

Subsystem changes:
	Air:
		Added air to the dynamic wait subsystem. upper bound: 50, lower bound: 5, cost delta: 3 times process cost
		Air now fires 4 times faster when it can do so without lagging things up
		Pipenet has been merged into air
		Atmos machinery now processes with process_atmos(), ticked by air, not machinery.
		Hotspots (the fire object) are now object pooled
	Pipenet:
		Deleted, added to air
	Machinery:
		Moved all atmos calcualtions in all objects's process() to process_atmos().
	Lighting:
		Added Lighting to the dynamic wait subsystem. upper bound: 20, lower bound: 5, cost delta: 3 times process cost
	Ticker:
		Fixed ticker not updating the lobby panel when game start delayed
		Fixed the game start timer updating rapidly from queued fires when game start delay is removed
	Garbage/qdel:
		qdel will now limit its process time to 2ds a fire.
		qdel can now be given hints as a return to Destroy() as to what should be done with the object.
		the options are:
			queue: (default) this is the normal behavior.
			letmelive: old default to non-null/zero. does nothing with the object
			iwillgc: functionally the same as above, mainly to let people working with objects know that the object will not be queued for GC checking
			harddel: this will queue the object to be deleted without storing a soft reference, mainly to save locate() processing time.
			harddel_now: this will del() the object. To allow for a clean removal of every del() not in qdel
		All objects have been updated to the new system, harddel and iwillgc was not added to any new objects.
		Fixed some objects not GCing because they didn't properlly clear references in Destory()
		Fixed some objects getting qdel'ed preventing other objects from getting GCed because they did not null their reference to that object.

MrsTonedOne committed:
Fixes a minor goof in powernets from fixing merge conflicts

xxalpha commented:
:+1: for faster pipes

RemieRichards commented on code/_onclick/hud/alert.dm:
```diff
@@ -200,5 +200,4 @@ Reharging stations are available in robotics, the dormitory's bathrooms. and the
 		return usr.client.Click(master, location, control, params)
 
 /obj/screen/alert/Destroy()
-	PlaceInPool(src)
```
Did you forget to add ```return QDEL_HINT_PUTINPOOL```?

MrStonedOne replied:
That is odd. ya, I did. hmm, Fixing.

I did most of the destroy review/rework while drunk, high, and running on fumes.

vista- replied:
>PUTINPOOL

russia has invaded

RemieRichards replied:
@MrStonedOne you like to live dangerously then :P

MrStonedOne replied:
ok, what the fuck, it is there on my copy.

must have been a rebase goof.

MrStonedOne replied:
What the fuck git.

If you click view whole file. Its there.

RemieRichards commented:
:+1: Great work. Travis did throw a shit fit though, lots of missing variables.

MrsTonedOne committed:
Makes hud/alert's destroy request obj pooling

MrsTonedOne committed:
Rebase mishap fix part 1

MrsTonedOne committed:
git rebase mishap fix part 2

MrStonedOne commented on code/modules/admin/admin.dm:
```diff
@@ -637,8 +638,8 @@ var/global/floorIsLava = 0
-	ticker.can_fire = !ticker.can_fire
-	if(!ticker.can_fire)
+	ticker.delay_start = !ticker.delay_start
+	if(!ticker.delay_start)
```
oh hey, i need to invert this check

MrsTonedOne committed:
Fixes delay start giving the wrong message

Alek2ander commented:
:gem: Double praise :gem:

Aranclanos commented:
three things, two extremly minor and not really important and a complaint:
-process_atmos_mechinery() not using var/thing and also checking if the machine exists (if(M))
-setup_allturfs() contains pipenet machinery settup
-making atmos 4 times faster with no real improvement is retarded

Anonus commented:
One thing; atmos is already four times faster on the live servers, this just means atmos can slow down if it'd be too much of a drain.

RemieRichards commented:
@Aranclanos the improvement is that the system can now slow itself down it it's got to do lots of calculations, which means on average it can be 4 times faster with "normal" amounts of calculations.

Aranclanos commented:
that's lovely, but the system isn't made to be run like that

Fox-McCloud commented:
Still don't see why you guys don't just port the process scheduler.

MrStonedOne commented on code/controllers/subsystem/air.dm:
```diff
@@ -124,11 +196,8 @@ var/datum/subsystem/air/SSair
-		if(EG.breakdown_cooldown > 20)
-			EG.dismantle()
\ No newline at end of file
+/datum/subsystem/air/proc/setup_atmos_machinery(z_level)
```
So the diffs are showing something fucking weird.

This is not in setup all turfs. this is in its own proc called setup_atmos_machinery. If you click view file, thats what you see. no idea why github is showing this diff so odd. and more importantly, no idea what this file will look like if this pr gets merged. what the fuck git.

MrStonedOne commented:
@Aranclanos 
>-process_atmos_machinery() not using var/thing and also checking if the machine exists (if(M))
I have it doing it exactly as the machinery controller does it, minus the power use part.
>-setup_allturfs() contains pipenet machinery setup
It actually doesn't, something odd as hell happened with the rebase. Click view file to see how that actually is.
>-making atmos 4 times faster with no real improvement is retarded
Its what the people want. and object pooling hotspots has made the lag issue less of an issue.

MrsTonedOne committed:
Minor rework of lobby admin verbs

Both me and #9228 fixed the same bug, so I reverted mine

Starting the round with the start now verb will now show a tip of the round.

MrsTonedOne committed:
pr-9231 changelog and minor subsystem tweaks

Aranclanos commented:
"its what people want" is not a good answer, the speed var was a gimmick, not something you could change  with the server configurations, if you want to do something, put some effort and do it properly instead of taking a huge dump on the code.
Linda is not made to do this, this shouldn't be merged for that exact reason.

HoaxIT commented:
:+1: for not merging this as @Aranclanos already said, Linda is not made to do this.

Also, as @Fox-McCloud already said, `Still don't see why you guys don't just port the process scheduler.` we're currently running it at hippie, and it works pretty okay.

optimumtact commented:
Because nobody wants to port it obviously

MrStonedOne commented:
@Aranclanos 
> if you want to do something, put some effort and do it properly instead of taking a huge dump on the code.

To put this in... oh what did you call it, "direct" terms you'd understand: Fuck off.

This is properly doing it. I'm sorry, but I see no reason to rewrite the entire atmos system when we have a perfectly good one right there.

I raised speed, examined the effects of doing so, fixed the parts of the system that didn't handle that well, and even went the extra effort to make the system automatically scale. I discovered issues with hotspots and fixed them, I went and tested every atmos machinery to ensure there wasn't odd behavior, found them (like cyro injecting chems too often, or heaters/coolers/portable atmospherics updating the dialog way too much) and fixed them.

I was working on this for 5 days. at least 3 of that was on the atmos controller and related things.

But no, I guess that wasn't "proper" enough for you. Maybe I should have just rewrote the entire system from scratch from the ground up, and lose the last 2 years+ of atmos/pipenets bug fixes, because you turned your nose up at the idea of somebody touching your precious.

>Linda is not made to do this, this shouldn't be merged for that exact reason.

[citation needed]

Just because you didn't design it to run at 5ds, doesn't mean it can't run at 5ds. Can you even name one issue with doing this? Because it seems like your solution is "It can be laggy in certain circumstances, so rather than have it dynamically scale how much it processes we should just have it only run at insanely slow rates." And honestly, I ain't hearing it, thats shit logic.

MrsTonedOne committed:
Merge remote-tracking branch 'upstream/master' into step-aside-aran-im-here-to-touch-your-precious

Conflicts:
	code/controllers/subsystem/ticker.dm

Aranclanos commented:
It doesn't matter how much time and work you put on this, merging the pipenets into the air datum isn't "touching linda" (which could have been avoided by just using a simple global var that they could share to have the same timers, kinda silly)
I'm all the citation you'll ever need, but in terms on proper maintaining, you actually need to get your hands messy with how the turfs, groups and gax mixtures handle to make it run at a faster speed, to avoid overheads. You know, put effort.
also go fug uself :DDD

MrStonedOne commented:
So, Let me see if I understand this right.

Because you have some absurd obsession with pipenets processing separately in their own controller, you think I should add ANOTHER snowflake global variable, and snowflake up the mastercontroller, just so two subsystems can process in tune? When it's a much simpler solution to just merge all atmos processing into one subsystem. Not to mention that your solution wouldn't even take into account the issue that under this proposed system, air's cost wouldn't take into account pipenet's processing time when deciding how to scale, so now master controller needs even more snowflaking.

For what? What exactly is so damn important about keeping pipenet's subsystem separate from air? What is gained by doing that? What is lost by doing my solution?

>I'm all the citation you'll ever need, but in terms on proper maintaining, you actually need to get your hands messy with how the turfs, groups and gax mixtures handle to make it run at a faster speed

But it already runs at a fast enough speed. That would be something I might look at doing if I wanted to drop atmo's lower bounds to 1ds. It handled having every door and window getting mass deleted just fucking fine at 5ds, and yes, I verified that the turfs got activated

I've done the testing. Both on my computer, and on the live servers when I edited atmo's processing rate to 5ds. I know what linda can handle.

Fox-McCloud commented:
@Aranclanos 

Don't worry Aran--we're attempting to remain as true to your system as possible =p

optimumtact commented:
Status Btfo [X]
